File: Medium/kth_Smallest_Element_In_A_BST/kthSmallestElementInABST.js
// Given a binary search tree, write a function kthSmallest to find the kth smallest element in it.

// Note: 
// You may assume k is always valid, 1 ≤ k ≤ BST's total elements.

// Example 1:
// Input: root = [3,1,4,null,2], k = 1
//    3
//   / \
//  1   4
//   \
//    2
// Output: 1

// Example 2:
// Input: root = [5,3,6,2,4,null,null,1], k = 3
//        5
//       / \
//      3   6
//     / \
//    2   4
//   /
//  1
// Output: 3

// Follow up:
// What if the BST is modified (insert/delete operations) often and you need to find the kth smallest frequently? How would you optimize the kthSmallest routine?


/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function(root, k) {
  if(!root) return null;
  return findNode(root, k);
};

const findNode = (root, k) => {
  const count = countNodes(root.left);
  if(k <= count) return findNode(root.left, k);
  else if(k > count + 1) return findNode(root.right, k - count - 1);
  return root.val;
};

const countNodes = (root) => {
  return root ? 1 + countNodes(root.left) + countNodes(root.right) : 0;
};

File: Medium/kth_Smallest_Element_In_A_BST/kthSmallestElementInABST.cs
// Given a binary search tree, write a function kthSmallest to find the kth smallest element in it.

// Note: 
// You may assume k is always valid, 1 ≤ k ≤ BST's total elements.

// Example 1:
// Input: root = [3,1,4,null,2], k = 1
//    3
//   / \
//  1   4
//   \
//    2
// Output: 1

// Example 2:
// Input: root = [5,3,6,2,4,null,null,1], k = 3
//        5
//       / \
//      3   6
//     / \
//    2   4
//   /
//  1
// Output: 3

// Follow up:
// What if the BST is modified (insert/delete operations) often and you need to find the kth smallest frequently? How would you optimize the kthSmallest routine?

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
  private int val;
  private int count;
  
  public int KthSmallest(TreeNode root, int k) {
    count = 0;
    FindNode(root, k);
    return val;
  }
  
  private void FindNode(TreeNode root, int k) {
    if(root == null) return;
    
    FindNode(root.left, k);
    
    count++;    
    if(count == k) {
      val = root.val;
      return;
    }
    
    FindNode(root.right, k);
  }
}